### DISCUSSION TITLE
CBOR serialization and removing RFC7049#2.4 type hints

### DISCUSSION THREAD

**Discussion by lachlan-ibm** [FasterXML/jackson-dataformats-binary#516] [Q&A] [answered]
I am working with a CBOR data structure which I am attempting to serialize from a HashMap. 

I have specific requirements on the CBOR encoding, the requirement I am having troubles with is 
related to https://datatracker.ietf.org/doc/html/rfc7049#section-2.4

The requirement I have is that CBOR I generate cannot have any item tags (as defined in section 2.4 of RFC7049)

Currently the CBOR I am generating has the following structure:
hex: a421c24301000120c2590101008e3497499dea33dd45d90eb866b01f9ebccb73e273472b4f9ab2e23b58aca3538daa2a2440af8ef0364b33f971a3d828a37faf25c01329233ce325d4048f2906e7e3e1f00ef1bd80ed2b69c565aa3fa475e131ebe72c5ae8dd2ce063e87ca644926e160b5d356ffe7cf37925e5306475aeb2ab5ce3ad8450091dda2f91b379e7fa5ac2e2f91751a0d270899e954259aa63d8df2a1d138ca04b63f735f539370b96a3227b345d5366a18f87c22d60ee28c475f9b194563b007d5fb8c68e79516d051e62635dfc567acf57da0e1a474e776297efa1fff61bdffa11dd4da9b35133d22596de0a96e4be6521f8b1cc9e2909e027ff0545781ce60eb3917c38aea0a7010303390100

parsed:
A4                                      # map(4)
   21                                   # negative(1)
   C2                                   # tag(2)
      43                                # bytes(3)
         010001                         # "\u0001\u0000\u0001"
   20                                   # negative(0)
   C2                                   # tag(2)
      59 0101                           # bytes(257)
         008E3497499DEA33DD45D90EB866B01F9EBCCB73E273472B4F9AB2E23B58ACA3538DAA2A2440AF8EF0364B33F971A3D828A37FAF25C01329233CE325D4048F2906E7E3E1F00EF1BD80ED2B69C565AA3FA475E131EBE72C5AE8DD2CE063E87CA644926E160B5D356FFE7CF37925E5306475AEB2AB5CE3AD8450091DDA2F91B379E7FA5AC2E2F91751A0D270899E954259AA63D8DF2A1D138CA04B63F735F539370B96A3227B345D5366A18F87C22D60EE28C475F9B194563B007D5FB8C68E79516D051E62635DFC567ACF57DA0E1A474E776297EFA1FFF61BDFFA11DD4DA9B35133D22596DE0A96E4BE6521F8B1CC9E2909E027FF0545781CE60EB3917C38AEA0A7 # "\u0000\x8E4\x97I\x9D\xEA3\xDDE\xD9\u000E\xB8f\xB0\u001F\x9E\xBC\xCBs\xE2sG+O\x9A\xB2\xE2;X\xAC\xA3S\x8D\xAA*$@\xAF\x8E\xF06K3\xF9q\xA3\xD8(\xA3\u007F\xAF%\xC0\u0013)#<\xE3%\xD4\u0004\x8F)\u0006\xE7\xE3\xE1\xF0\u000E\xF1\xBD\x80\xED+i\xC5e\xAA?\xA4u\xE11\xEB\xE7,Z\xE8\xDD,\xE0c\xE8|\xA6D\x92n\u0016\v]5o\xFE|\xF3y%\xE50du\xAE\xB2\xAB\\㭄P\t\u001D\xDA/\x91\xB3y\xE7\xFAZ\xC2\xE2\xF9\u0017Q\xA0\xD2p\x89\x9E\x95BY\xAAc\xD8\xDF*\u001D\u0013\x8C\xA0Kc\xF75\xF597\v\x96\xA3\"{4]Sf\xA1\x8F\x87\xC2-`\xEE(\xC4u\xF9\xB1\x94V;\u0000}_\xB8ƎyQm\u0005\u001Ebc]\xFCVz\xCFW\xDA\u000E\u001AGNwb\x97\xEF\xA1\xFF\xF6\e\xDF\xFA\u0011\xDDM\xA9\xB3Q3\xD2%\x96\xDE\n\x96\xE4\xBEe!\xF8\xB1̞)\t\xE0'\xFF\u0005Ex\u001C\xE6\u000E\xB3\x91|8\xAE\xA0\xA7"
   01                                   # unsigned(1)
   03                                   # unsigned(3)
   03                                   # unsigned(3)
   39 0100                              # negative(256)
   
   
note the "tag(2)" entries which are a type hint for the byte array values in the map.

The output I need is as follows:
hex:
a4214301000120590101008e3497499dea33dd45d90eb866b01f9ebccb73e273472b4f9ab2e23b58aca3538daa2a2440af8ef0364b33f971a3d828a37faf25c01329233ce325d4048f2906e7e3e1f00ef1bd80ed2b69c565aa3fa475e131ebe72c5ae8dd2ce063e87ca644926e160b5d356ffe7cf37925e5306475aeb2ab5ce3ad8450091dda2f91b379e7fa5ac2e2f91751a0d270899e954259aa63d8df2a1d138ca04b63f735f539370b96a3227b345d5366a18f87c22d60ee28c475f9b194563b007d5fb8c68e79516d051e62635dfc567acf57da0e1a474e776297efa1fff61bdffa11dd4da9b35133d22596de0a96e4be6521f8b1cc9e2909e027ff0545781ce60eb3917c38aea0a7010303390100

parsed:
A4                                      # map(4)
   21                                   # negative(1)
   43                                   # bytes(3)
      010001                            # "\u0001\u0000\u0001"
   20                                   # negative(0)
   59 0101                              # bytes(257)
      008E3497499DEA33DD45D90EB866B01F9EBCCB73E273472B4F9AB2E23B58ACA3538DAA2A2440AF8EF0364B33F971A3D828A37FAF25C01329233CE325D4048F2906E7E3E1F00EF1BD80ED2B69C565AA3FA475E131EBE72C5AE8DD2CE063E87CA644926E160B5D356FFE7CF37925E5306475AEB2AB5CE3AD8450091DDA2F91B379E7FA5AC2E2F91751A0D270899E954259AA63D8DF2A1D138CA04B63F735F539370B96A3227B345D5366A18F87C22D60EE28C475F9B194563B007D5FB8C68E79516D051E62635DFC567ACF57DA0E1A474E776297EFA1FFF61BDFFA11DD4DA9B35133D22596DE0A96E4BE6521F8B1CC9E2909E027FF0545781CE60EB3917C38AEA0A7 # "\u0000\x8E4\x97I\x9D\xEA3\xDDE\xD9\u000E\xB8f\xB0\u001F\x9E\xBC\xCBs\xE2sG+O\x9A\xB2\xE2;X\xAC\xA3S\x8D\xAA*$@\xAF\x8E\xF06K3\xF9q\xA3\xD8(\xA3\u007F\xAF%\xC0\u0013)#<\xE3%\xD4\u0004\x8F)\u0006\xE7\xE3\xE1\xF0\u000E\xF1\xBD\x80\xED+i\xC5e\xAA?\xA4u\xE11\xEB\xE7,Z\xE8\xDD,\xE0c\xE8|\xA6D\x92n\u0016\v]5o\xFE|\xF3y%\xE50du\xAE\xB2\xAB\\㭄P\t\u001D\xDA/\x91\xB3y\xE7\xFAZ\xC2\xE2\xF9\u0017Q\xA0\xD2p\x89\x9E\x95BY\xAAc\xD8\xDF*\u001D\u0013\x8C\xA0Kc\xF75\xF597\v\x96\xA3\"{4]Sf\xA1\x8F\x87\xC2-`\xEE(\xC4u\xF9\xB1\x94V;\u0000}_\xB8ƎyQm\u0005\u001Ebc]\xFCVz\xCFW\xDA\u000E\u001AGNwb\x97\xEF\xA1\xFF\xF6\e\xDF\xFA\u0011\xDDM\xA9\xB3Q3\xD2%\x96\xDE\n\x96\xE4\xBEe!\xF8\xB1̞)\t\xE0'\xFF\u0005Ex\u001C\xE6\u000E\xB3\x91|8\xAE\xA0\xA7"
   01                                   # unsigned(1)
   03                                   # unsigned(3)
   03                                   # unsigned(3)
   39 0100                              # negative(256)

Is there a Jackson class I can override to stop generating these tags?

I have been trying to override ArraySerializerBase.java and MapEntrySerializer.java with no success

**Comment by cowtowncoder** [comment]
Ok this is wrong place: CBOR is under `jackson-dataformats-binary`. But need to enable discussion there (should be added as an issue but no option to "convert-and-transfer") first, then transfer.

**Comment by cowtowncoder** [comment, marked as answer]
Ok so, no, you cannot force this easily from databind level (where JsonSerializers are defined) which is format-agnostic. None of serializers know of actual underlying format, or encoding details.

So ability to suppress type tags/hints would need to be done at streaming API level, probably by adding

    CBORGenerator.Feature.FEATURE_NAME

(where FEATURE_NAME would be like `USE_TYPE_TAGS` or `USE_TYPE_TAG_xxx`)

and then making `CBORGenerator` take that into account when encoding.

Above is my initial thinking without considering any of the details (and only vaguely remembering how type tags / hints were used).

**Reply by lachlan-ibm** [reply]
Thank you for the response and advice.
